quickwit-metastore/src/metastore/postgresql_metastore.rs

```diff
@@ -0,0 +1,1120 @@
+                .collect::<Vec<String>>(),
+            split_metadata_json: split_metadata_and_footer_offsets_json,
+            index_id: index_id.to_string(),
+        };
+
+        conn.transaction::<_, diesel::result::Error, _>(|| {
+            // Insert a new split metadata as `Staged` state.
+            let insert_staged_split_statement =
+                diesel::insert_into(schema::splits::dsl::splits).values(&model_split);
+            debug!(sql=%debug_query::<Pg, _>(&insert_staged_split_statement).to_string());
+            insert_staged_split_statement.execute(&*conn)?;
+
+            Ok(())
+        })
+        .map_err(|err| match err {
+            DatabaseError(kind, err_info) => match kind {
+                DatabaseErrorKind::ForeignKeyViolation => MetastoreError::IndexDoesNotExist {
+                    index_id: index_id.to_string(),
+                },
+                _ => MetastoreError::InternalError {
+                    message: "Failed to stage split".to_string(),
+                    cause: anyhow::anyhow!(err_info.message().to_string()),
+                },
+            },
+            _ => MetastoreError::InternalError {
+                message: "Failed to stage split".to_string(),
+                cause: anyhow::anyhow!(err),
+            },
+        })?;
+
+        Ok(())
+    }
+
+    async fn publish_splits<'a>(
+        &self,
+        index_id: &str,
+        split_ids: &[&'a str],
+        checkpoint_delta: CheckpointDelta,
+    ) -> MetastoreResult<()> {
+        let conn = self
+            .connection_pool
+            .get()
+            .map_err(|err| MetastoreError::InternalError {
+                message: "Failed to get connection".to_string(),
+                cause: anyhow::anyhow!(err),
+            })?;
+
+        // Get index metadata.
+        let select_index_statement =
+            schema::indexes::dsl::indexes.filter(schema::indexes::dsl::index_id.eq(index_id));
+        debug!(sql=%debug_query::<Pg, _>(&select_index_statement).to_string());
+        let model_index = select_index_statement
+            .first::<model::Index>(&conn)
+            .map_err(|err| match err {
+                diesel::result::Error::NotFound => MetastoreError::IndexDoesNotExist {
+                    index_id: index_id.to_string(),
+                },
+                _ => MetastoreError::InternalError {
+                    message: "Failed to select index".to_string(),
+                    cause: anyhow::anyhow!(err),
+                },
+            })?;
+
+        // Deserialize the checkpoint from the database model.
+        let mut index_metadata =
+            model_index
+                .make_index_metadata()
+                .map_err(|err| MetastoreError::InternalError {
+                    message: "Failed to deserialize index metadata".to_string(),
+                    cause: anyhow::anyhow!(err),
+                })?;
+
+        // Apply checkpoint_delta
+        index_metadata
```

Comment:
I'd rather have the checkpoint check logic to be part of the transaction...

Reply:
If I move the checkpoint logic to a transaction, I would expect to roll back the transaction if an error occurs in `Checkpoint::try_apply_delta`.
In `transaction` functions, it seems that if I want to roll back with user code other than DB operations, I have to return `diesel::result::Error:.RollbackTransaction`.
Then the `MetastoreError::IncompatibleCheckpointDelta` that the `publish_splits` function has to return cannot be known outside the transaction.
Is there a good way to know non-diesel::result::Error that occurs in a transaction outside of the transaction?